[BOAT] auto raid moderation

BenSegal855 opened the pull request:
This PR will kick any user who joined the guild less than one minute prior and sent the same message three times. It will ban any user who does this twice. All messages sent by any punished raider is also deleted (it even uses bulk deletes when possible to avoid api spam)

BenSegal855 committed:
feat: raid auto moderation

BenSegal855 committed:
feat: raid messages are deleted on kick/ban

cyyynthia commented on boat/src/modules/raidmod.ts:
```diff
@@ -0,0 +1,117 @@
+import config from '../config.js';
+import { ban, kick } from '../mod.js';
+
+class RaidMessage {
```
Couldn't this be simplified to a plain object created in `addRaidMessage`?

cyyynthia commented on boat/src/modules/raidmod.ts:
```diff
@@ -0,0 +1,117 @@
+// Any new members who send more than THRESHOLD messages with the same content, will be kicked
+const THRESHOLD = 2
+const raiderBuffer = new Map<string, number>()
+const messageBuffer = new Collection(RaidMessage)
```
Why use a collection from Eris here? This could easily be replaced by a vanilla Map/Set, or I'm missing something?

cyyynthia commented on boat/src/modules/raidmod.ts:
```diff
@@ -0,0 +1,117 @@
+const messageBuffer = new Collection(RaidMessage)
+
+async function process(this: CommandClient, msg: Message<GuildTextableChannel>) {
+  if (msg.guildID !== config.discord.ids.serverId || !msg.member || msg.member.joinedAt < Date.now() - 60e3) return;
```
A minute is rather short, and since the source is public people will just wait before spamming (because we know from which very wholesome community some raids were coming from) <!-- vizality cough cough -->

Maybe increase the limit to at least 10mins (to match Discord's moderation setting), or even higher like a day. It might be a good idea to always catch accounts older than a month to make use of burner accounts more annoying.

cyyynthia commented on boat/src/modules/raidmod.ts:
```diff
@@ -0,0 +1,117 @@
+
+  addRaidMessage(msg)
+
+  if (isRaider(msg.author.id, msg.content)) {
```
It may be a good idea when we detect an incoming wave of raiders to automatically set the server moderation level to highest for a set period of time to help slow down the wave.

cyyynthia replied:
Implementing this should be fairly easy, as it's basically a timer: if in the last let's say 5mins we catch 5 people, call [Guild#edit](https://abal.moe/Eris/docs/Guild#function-edit) and schedule a cleanup task which sets it back to what it was later

BenSegal855 replied:
Resolved by [397e119](https://github.com/powercord-org/powercord-backend/pull/69/commits/397e119eaea79a926857df1446d42790e04cdd54)

cyyynthia commented on boat/src/modules/raidmod.ts:
```diff
@@ -0,0 +1,117 @@
+
+  if (isRaider(msg.author.id, msg.content)) {
+    if (await this.mongo.collection('raiders').countDocuments({ userId: msg.author.id }) > 0) {
+      deleteRaiderMessages(this, msg.author.id)
+      ban(msg.channel.guild, msg.author.id, this.user, 'Repeat raider')
```
You can tell Discord to delete messages along with the ban, to avoid having to do it yourself.

cyyynthia commented on boat/src/modules/raidmod.ts:
```diff
@@ -0,0 +1,117 @@
+      deleteRaiderMessages(this, msg.author.id)
+      ban(msg.channel.guild, msg.author.id, this.user, 'Repeat raider')
+    } else {
+      deleteRaiderMessages(this, msg.author.id)
+      kick(msg.channel.guild, msg.author.id, this.user, 'Detected raid spam')
```
Same as above, except that here you'd do a softban (ban & unban right after). Must come with proper flag in the reason field so the mod logger can mark the ban as a kick and ignore the unban event (a `[Softban]` flag should do the trick I think?)

cyyynthia commented on boat/src/modules/raidmod.ts:
```diff
@@ -0,0 +1,117 @@
+
+function isRaider(user: string, message: string): boolean {
+  const raiderHash = createHash('sha256').update(`${user}${message}`).digest('base64').toString()
+  let count = raiderBuffer.get(raiderHash)
+
+  if (count === undefined) {
+    raiderBuffer.set(raiderHash, 1)
+    setTimeout(() => removeRaider(raiderHash), 10e3);
+    return false;
+  }
```
```suggestion
  let count = raiderBuffer.get(raiderHash) ?? 0
```

cyyynthia commented on boat/src/modules/raidmod.ts:
```diff
@@ -0,0 +1,117 @@
+}
+
+function isRaider(user: string, message: string): boolean {
+  const raiderHash = createHash('sha256').update(`${user}${message}`).digest('base64').toString()
```
Why use a hash here? I'm assuming it's to save on memory (and to store just 32 bytes instead of up to 2018 chars), but then you could just use sha1 and compare buffers instead, I think?

cyyynthia reviewed (changes requested):
There's in addition to the comments I left also a fair chunk of code style issues (no semi, no `undefined` (use `void 0` instead), ...).

BenSegal855 committed:
tweak: new user detection and "linting"

BenSegal855 committed:
tweak: use ban deletes

BenSegal855 committed:
fix: monitor new members

BenSegal855 committed:
tweak: Smarter inital count

Co-authored-by: Cynthia cynthia@cynthia.dev>

BenSegal855 committed:
feat: softban function & command

BenSegal855 committed:
tweak: switch to sha1

BenSegal855 committed:
tweak: Switch faid kick to softbans

BenSegal855 committed:
chore: Add jsDocs to util and mod functions

cyyynthia commented on boat/src/commands/mod/softban.ts:
```diff
@@ -0,0 +1,36 @@
+import { isStaff } from '../../util.js'
+import config from '../../config.js'
+
+const USAGE_STR = `Usage: ${config.discord.prefix}ban <mention || id> [reason]`
```
```suggestion
const USAGE_STR = `Usage: ${config.discord.prefix}softban <mention || id> [reason]`
```

cyyynthia commented on boat/src/commands/mod/softban.ts:
```diff
@@ -0,0 +1,36 @@
+import type { GuildTextableChannel, Message } from 'eris'
```
Missing license header

cyyynthia commented on boat/src/modules/modlog.ts:
```diff
@@ -67,12 +67,19 @@ function processBanFactory (type: 'add' | 'remove'): (guild: Guild, user: User)
+    const soft = reason.startsWith('[soft]')
+    if (soft) {
+      if (type === 'remove') return
+      reason = reason.replace('[soft]', '')
```
```suggestion
      reason = reason.replace('[soft] ', '')
```

BenSegal855 committed:
Update boat/src/commands/mod/softban.ts

Co-authored-by: Cynthia <cynthia@cynthia.dev>

BenSegal855 committed:
Update boat/src/modules/modlog.ts

Co-authored-by: Cynthia <cynthia@cynthia.dev>

BenSegal855 committed:
tweak: add license header

BenSegal855 committed:
feat: raid mode
- staff commands to enter & exit raid mode
- auto raid detection enters rade mode
- raid task to exit raid mode automaticly

cyyynthia commented on boat/src/commands/mod/endraid.ts:
```diff
@@ -0,0 +1,40 @@
+import { exitRaidMode, getRaidStatus } from '../../raidMode.js'
+import { isStaff, } from '../../util.js'
+
+export async function executor (msg: Message<GuildTextableChannel>,): Promise<Message | void> {
```
```suggestion
export function executor (msg: Message<GuildTextableChannel>): void {
```

cyyynthia commented on boat/src/commands/mod/endraid.ts:
```diff
@@ -0,0 +1,40 @@
+export async function executor (msg: Message<GuildTextableChannel>,): Promise<Message | void> {
+  if (!msg.member) return // ???
+  if (!isStaff(msg.member)) {
+    return msg.channel.createMessage('no')
```
```suggestion
    msg.channel.createMessage('no')
    return
```

cyyynthia commented on boat/src/commands/mod/endraid.ts:
```diff
@@ -0,0 +1,40 @@
+  }
+
+  if (!getRaidStatus()) {
+    return msg.channel.createMessage('Raid mode is not currently active.')
```
```suggestion
    msg.channel.createMessage('Raid mode is not currently active.')
    return
```

cyyynthia commented on boat/src/commands/mod/endraid.ts:
```diff
@@ -0,0 +1,40 @@
+    return msg.channel.createMessage('Raid mode is not currently active.')
+  }
+
+  exitRaidMode(msg.channel.guild, msg.author)
+
+  return msg.channel.createMessage('Raid mode disabled.')
```
```suggestion
  exitRaidMode(msg.channel.guild, msg.author)
  msg.channel.createMessage('Raid mode disabled.')
```

cyyynthia commented on boat/src/commands/mod/raid.ts:
```diff
@@ -0,0 +1,52 @@
+
+const USAGE_STR = `Usage: ${config.discord.prefix}raid <duration>`
+
+export async function executor (msg: Message<GuildTextableChannel>, [ rawDuration ]: [string]): Promise<Message | void> {
```
```suggestion
export function executor (msg: Message<GuildTextableChannel>, [ rawDuration ]: [ string ]): void {
```

cyyynthia commented on boat/src/commands/mod/raid.ts:
```diff
@@ -0,0 +1,52 @@
+export async function executor (msg: Message<GuildTextableChannel>, [ rawDuration ]: [string]): Promise<Message | void> {
+  if (!msg.member) return // ???
+  if (!isStaff(msg.member)) {
+    return msg.channel.createMessage('no')
```
```suggestion
    msg.channel.createMessage('no')
    return
```

cyyynthia commented on boat/src/commands/mod/raid.ts:
```diff
@@ -0,0 +1,52 @@
+  }
+
+  if (getRaidStatus()) {
+    return msg.channel.createMessage('Raid mode is currently active.')
```
```suggestion
    msg.channel.createMessage('Raid mode is currently active.')
    return
```

cyyynthia commented on boat/src/commands/mod/raid.ts:
```diff
@@ -0,0 +1,52 @@
+  }
+
+  if (!rawDuration) {
+    return msg.channel.createMessage(USAGE_STR)
```
```suggestion
    msg.channel.createMessage(USAGE_STR)
    return
```

cyyynthia commented on boat/src/commands/mod/raid.ts:
```diff
@@ -0,0 +1,52 @@
+
+  const duration = parseDuration(rawDuration)
+  if (!duration) {
+    return msg.channel.createMessage('Invalid duration')
```
```suggestion
    msg.channel.createMessage('Invalid duration')
    return
```

cyyynthia commented on boat/src/commands/mod/raid.ts:
```diff
@@ -0,0 +1,52 @@
+    return msg.channel.createMessage('Invalid duration')
+  }
+
+  enterRaidMode(msg.channel.guild, msg.author, duration)
+
+  return msg.channel.createMessage('Raid mode activated.')
```
```suggestion
  enterRaidMode(msg.channel.guild, msg.author, duration)
  msg.channel.createMessage('Raid mode activated.')
```

cyyynthia commented on boat/src/commands/mod/raid.ts:
```diff
@@ -0,0 +1,52 @@
+ */
+
+import { Message, GuildTextableChannel } from 'eris'
+import config from '../../config.js'
```
Import config last (not a linting rule or anything, just a preference)

cyyynthia commented on boat/src/index.ts:
```diff
@@ -29,6 +29,7 @@ import { readdirRecursive } from './util.js'
 import { loadLaws } from './laws.js'
 import { loadBlacklist } from './blacklistCache.js'
 import config from './config.js'
```
Prefer importing config last

cyyynthia commented on boat/src/modules/raidmod.ts:
```diff
@@ -0,0 +1,88 @@
+
+import { createHash } from 'crypto'
+import { CommandClient, GuildTextableChannel, Message } from 'eris'
+import config from '../config.js'
```
Prefer importing config last

cyyynthia commented on boat/src/modules/raidmod.ts:
```diff
@@ -0,0 +1,88 @@
+const raiderBuffer = new Map<string, number>()
+const DAY_MS = 24 * 36e5
+
+let activeRaiders = 0;
```
```suggestion
let activeRaiders = 0
```

cyyynthia commented on boat/src/raidMode.ts:
```diff
@@ -0,0 +1,84 @@
+ */
+
+import { CommandClient, Guild, TextChannel, User } from 'eris'
+import config from './config.js'
```
Prefer importing config last

cyyynthia commented on boat/src/modules/raidmod.ts:
```diff
@@ -0,0 +1,88 @@
+}
+
+function removeRaider(hash: string) {
+  let count = raiderBuffer.get(hash)
+
+  if (count === undefined) return
+
+  count--
+
+  if (count === 0) {
+    raiderBuffer.delete(hash)
+  } else {
+    raiderBuffer.set(hash, count)
+  }
```
```suggestion
  const count = raiderBuffer.get(hash)
  if (!count) return

  if (count === 1) {
    raiderBuffer.delete(hash)
  } else {
    raiderBuffer.set(hash, count - 1)
  }
```

cyyynthia commented on boat/src/raidMode.ts:
```diff
@@ -0,0 +1,84 @@
+/**
+ * Return a guild to the normal verification level.
+ * @param guild - the guild to have its verification level lowered
+ * @param mod - the moderator responsible fro lowering the verification level
```
```suggestion
 * @param mod - the moderator responsible for lowering the verification level
```

cyyynthia commented on boat/src/commands/mod/softban.ts:
```diff
@@ -0,0 +1,58 @@
+  }
+
+  softBan(msg.channel.guild, target, msg.author, reason, 1)
+  msg.channel.createMessage('yeeted')
```
```suggestion
  msg.channel.createMessage('yeeted & censored')
```

cyyynthia commented on boat/src/modules/modtasks.ts:
```diff
@@ -42,13 +43,24 @@ function processTasks (bot: CommandClient) {
+ * Schedule a task
+ * @param task - the task type
+ * @param guild - the guild the task will be operating in
+ * @param userId - the user the task will be operation on
```
```suggestion
 * @param userId - the user the task will be operated on
```

cyyynthia commented on boat/src/raidMode.ts:
```diff
@@ -0,0 +1,84 @@
+/**
+ * Raise a guilds verification level for a set amount of time.
+ * @param guild - the guild that will have its verification level raised
+ * @param mod - the moderator responsible for rasing the verification level
```
```suggestion
 * @param mod - the moderator responsible for raising the verification level
```

cyyynthia commented on boat/src/util.ts:
```diff
@@ -80,13 +106,24 @@ export function stringifyDiscordMessage (message: Message<GuildTextableChannel>)
+/**
+ * Convert a human readable duration to its numerical representation.
+ * @param duration - the human readable duration string
+ * @returns the numerical representation of `duration`
```
```suggestion
 * @returns the numerical representation of `duration` in ms
```

cyyynthia reviewed (changes requested):
A lot of nitpicking here (and I didn't note inconsistencies in where periods are placed :p)

BenSegal855 committed:
Update boat/src/commands/mod/endraid.ts

Co-authored-by: Cynthia <cynthia@cynthia.dev>

BenSegal855 committed:
Update boat/src/commands/mod/endraid.ts

Co-authored-by: Cynthia <cynthia@cynthia.dev>

BenSegal855 committed:
Update boat/src/commands/mod/endraid.ts

Co-authored-by: Cynthia <cynthia@cynthia.dev>

BenSegal855 committed:
Update boat/src/commands/mod/endraid.ts

Co-authored-by: Cynthia <cynthia@cynthia.dev>

BenSegal855 committed:
Update boat/src/commands/mod/raid.ts

Co-authored-by: Cynthia <cynthia@cynthia.dev>

BenSegal855 committed:
Update boat/src/commands/mod/raid.ts

Co-authored-by: Cynthia <cynthia@cynthia.dev>

BenSegal855 committed:
Update boat/src/commands/mod/raid.ts

Co-authored-by: Cynthia <cynthia@cynthia.dev>

BenSegal855 committed:
Update boat/src/commands/mod/raid.ts

Co-authored-by: Cynthia <cynthia@cynthia.dev>

BenSegal855 committed:
Update boat/src/commands/mod/raid.ts

Co-authored-by: Cynthia <cynthia@cynthia.dev>

BenSegal855 committed:
Update boat/src/commands/mod/raid.ts

Co-authored-by: Cynthia <cynthia@cynthia.dev>

BenSegal855 committed:
Update boat/src/modules/raidmod.ts

Co-authored-by: Cynthia <cynthia@cynthia.dev>

BenSegal855 committed:
Apply suggestions from code review

Just learned how to do batch these

Co-authored-by: Cynthia <cynthia@cynthia.dev>

BenSegal855 committed:
tweak: inport config last
